whisk.py

```diff
@@ -102,7 +102,25 @@ def parse_conf_file(path):
     # Re-parse, expanding variables
     env = os.environ.copy()
     env["WHISK_PROJECT_ROOT"] = project_root.absolute()
-    conf = yaml.load(ConfTemplate(conf_str).substitute(**env), Loader=yaml.Loader)
+    conf = yaml.load(conf_str, Loader=yaml.Loader)
+
+    # Recursively do environment substitution on all strings
+    def substitute(item):
+
+        if isinstance(item, dict):
+            for k in item:
+                item[k] = substitute(item[k])
+            return item
+        elif isinstance(item, list):
+            for i in range(0, len(item)):
```

Comment:
I deliberately used range() here even though that's not idiomatic Python loop iteration. We don't want to disturb the ordering of elements in the list or replace the overall list instance, so doing a comprehension wouldn't be appropriate.